### DISCUSSION TITLE
How to wait for consumer to complete poll function after stop

### DISCUSSION THREAD

**Discussion by mikicho** [bbc/sqs-consumer#410] [Feature Request]
### Background

TL;DR:
The `start` function is sync but leaves "open" promise that users can't `await` for.

Example:

I have a consumer that prints a log on the `message_processed` event:
```
consumer.on('message_processed', (message) => {
    console.log('Yay');
});
```

The problem is that my tests (and the process in production) are completed before the listener gets executed because the `stop` function doesn't wait for `poll` function to complete.

### Objectives

`stop` function should wait for the poll function to complete.

OR

Emit `shutdown` event only after the `poll` function is complete instead of inside the `stop` function. so users will be able to await for `shutdown` event

**Comment by mikicho** [comment]
It may somewhat related to https://github.com/bbc/sqs-consumer/discussions/392

**Comment by nicholasgriffintn** [comment]
Hey! Sorry for the late reply.

This is something that we have had raised before and we have noted it as a future exploration task, due to how Consumer works though, this is slightly tricky.

**Reply by mikicho** [reply]
Why can't we emit shutdown/drained event?

**Reply by nicholasgriffintn** [reply]
Because it's on a set timeout so effectively polling may sometimes run after stop, we do emit a stop event though.

**Reply by mikicho** [reply]
This is why I think we should emit another event when polling actually stops:
e.g. here: https://github.com/bbc/sqs-consumer/blob/main/src/consumer.ts#L178

**Reply by nicholasgriffintn** [reply]
Yeah so currently there's no way of actually knowing how if there's a soon to be active timer, they should be cancelled on stop, but this is there to catch the cases where sometimes they don't get cancelled.

We'd have to rewrite the queuing in order to know if there's anything in it or not.

**Reply by mikicho** [reply]
Oh.. are you worried about inflight messages?

**Reply by nicholasgriffintn** [reply]
Yes that's basically the reason.

**Reply by mikicho** [reply]
Maybe we can keep a counter?

**Comment by barwin** [comment]
I've taken a stab at this with https://github.com/bbc/sqs-consumer/pull/454 -- There are a few different ways to go about this, so I'm happy to rework this PR with some guidance if maintainers are opposed to this approach.
